src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseBaseKernels.hpp

```diff
@@ -129,128 +132,353 @@ struct MobilityKernel
   }
 };
 
-/******************************** AccumulationKernel ********************************/
-struct AccumulationKernel
+/******************************** ElementBasedAssemblyKernel ********************************/
+
+/**
+ * @brief Internal struct to provide no-op defaults used in the inclusion
+ *   of lambda functions into kernel component functions.
+ * @struct NoOpFunc
+ */
+struct NoOpFunc
 {
+  template< typename ... Ts >
   GEOSX_HOST_DEVICE
-  GEOSX_FORCE_INLINE
-  static void
-  compute( real64 const & densNew,
-           real64 const & dens_n,
-           real64 const & dDens_dPres,
-           real64 const & poreVolNew,
-           real64 const & poreVol_n,
-           real64 const & dPoreVol_dPres,
-           real64 & localAccum,
-           real64 & localAccumJacobian )
+  constexpr void
+  operator()( Ts && ... ) const {}
+};
+
+/**
+ * @class ElementBasedAssemblyKernel
+ * @brief Define the interface for the assembly kernel in charge of accumulation
+ */
+template< typename SUBREGION_TYPE >
+class ElementBasedAssemblyKernel
+{
+
+public:
+
+  /**
+   * @brief Constructor
+   * @param[in] rankOffset the offset of my MPI rank
+   * @param[in] dofKey the string key to retrieve the degress of freedom numbers
+   * @param[in] subRegion the element subregion
+   * @param[in] fluid the fluid model
+   * @param[in] solid the solid model
+   * @param[inout] localMatrix the local CRS matrix
+   * @param[inout] localRhs the local right-hand side vector
+   */
+  ElementBasedAssemblyKernel( globalIndex const rankOffset,
+                              string const dofKey,
+                              SUBREGION_TYPE const & subRegion,
+                              constitutive::SingleFluidBase const & fluid,
+                              constitutive::CoupledSolidBase const & solid,
+                              CRSMatrixView< real64, globalIndex const > const & localMatrix,
+                              arrayView1d< real64 > const & localRhs )
+    :
+    m_rankOffset( rankOffset ),
+    m_dofNumber( subRegion.template getReference< array1d< globalIndex > >( dofKey ) ),
+    m_elemGhostRank( subRegion.ghostRank() ),
+    m_volume( subRegion.getElementVolume() ),
+    m_deltaVolume( subRegion.template getExtrinsicData< extrinsicMeshData::flow::deltaVolume >() ),
+    m_porosity_n( solid.getPorosity_n() ),
+    m_porosityNew( solid.getPorosity() ),
+    m_dPoro_dPres( solid.getDporosity_dPressure() ),
+    m_density_n( fluid.density_n() ),
+    m_density( fluid.density() ),
+    m_dDensity_dPres( fluid.dDensity_dPressure() ),
+    m_localMatrix( localMatrix ),
+    m_localRhs( localRhs )
+  {}
+
+  /**
+   * @struct StackVariables
+   * @brief Kernel variables (dof numbers, jacobian and residual) located on the stack
+   */
+  struct StackVariables
+  {
+public:
+
+    // Pore volume information
+
+    /// Pore volume at time n+1
+    real64 poreVolumeNew = 0.0;
```

Comment:
should we just call this `poreVolume` ?

Reply:
I modified the name to `poreVolume` here and in the compositional kernels